### PR TITLE
SPARK-12417. [SQL] Orc bloom filter options are not propagated during…

### PR DESCRIPTION
**Pull request by rajeshbalamohan**
Add option to have bloom filters in ORC write codepath.  Added changes to apply cleanly in master branch.

### CODE REVIEW

**Comment by rajeshbalamohan** [comment]
Closed #10375 which was the dup of this pull request. Review can be done on this pull request.  Thanks @JoshRosen

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcQuerySuite.scala [hunk 1]
```diff
@@ -394,4 +394,42 @@ class OrcQuerySuite extends QueryTest with BeforeAndAfterAll with OrcTest {
       }
     }
   }
+
+  test("SPARK-12417. Orc bloom filter options are not propagated during file " +
```
Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcQuerySuite.scala]: Change the title to

"SPARK-12417: Propagate Bloom filters when writing out ORC files"

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcQuerySuite.scala [hunk 2]
```diff
@@ -394,4 +394,42 @@ class OrcQuerySuite extends QueryTest with BeforeAndAfterAll with OrcTest {
...
+    "generation") {
+    withTempPath { dir =>
+      // Create separate sub dir for bloom filter testing
+      val path = new File(dir, "orc").getCanonicalPath
+
+      // Write some data
+      val data = (0 until 10).map { i =>
```
Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcQuerySuite.scala]: why do you need data to be in this form? can't it just be a list of numbers from 0 to 10?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcRelation.scala [hunk 3]
```diff
@@ -162,6 +162,10 @@ private[sql] class OrcRelation(
   extends HadoopFsRelation(maybePartitionSpec, parameters)
   with Logging {
 
+  /* org.apache.orc contains OrcConf which can be used later */
+  private[sql] val ORC_BLOOM_FILTER_COLUMNS = "orc.bloom.filter.columns"
```
Comment (rxin) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcRelation.scala]: it is really confusing to have ORC_BLOOM_FILTER_COLUMNS and orcBloomFilterCols in the same class.

I'd just inline the string in orcBloomFilterCols. Same for orcBloomFilterFpp.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcRelation.scala [hunk 4]
```diff
@@ -162,6 +162,10 @@ private[sql] class OrcRelation(
   extends HadoopFsRelation(maybePartitionSpec, parameters)
   with Logging {
 
+  /* org.apache.orc contains OrcConf which can be used later */
```
Comment (rxin) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcRelation.scala]: it's unclear what this message actually means - is this a todo for the future?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcQuerySuite.scala [hunk 5]
```diff
@@ -394,4 +394,42 @@ class OrcQuerySuite extends QueryTest with BeforeAndAfterAll with OrcTest {
...
+      createDataFrame(data).toDF("a", "b")
+        .write.option("orc.bloom.filter.columns", "*").orc(path)
+
+      // Verify if orc bloom filters are present. This can be verified via
+      // ORC RecordReaderImpl when it is made public. Until then, verify by
+      // dumping file statistics and checking whether bloom filter was added.
+      new File(path).listFiles().filter(_.getName.endsWith("orc")) map { file =>
+        withTempStream { buf =>
+          val fileDumpArgs = Array(file.getCanonicalPath)
+          FileDump.main(fileDumpArgs)
```
Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcQuerySuite.scala]: this seems really hacky and is depending on some non-public API.

Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcQuerySuite.scala]: Can't you just read the files directly?

**Pull request closed without merging**